lib/secretsscanner/scan/scan.go

```diff
@@ -0,0 +1,252 @@
+	}
+	return &Scanner{
+		dirs:        cfg.Dirs,
+		log:         cfg.Log,
+		privateKeys: make(map[string]*accessgraphsecretsv1pb.PrivateKey),
+	}, nil
+}
+
+// Scanner is a scanner that scans directories for secrets.
+type Scanner struct {
+	dirs []string
+	log  *slog.Logger
+	// privateKeys is a map of private keys found during the scan.
+	// The key is the path to the private key file and the value is the private key representation.
+	privateKeys map[string]*accessgraphsecretsv1pb.PrivateKey
+}
+
+// ScanPrivateKeys scans directories for SSH private keys.
+func (s *Scanner) ScanPrivateKeys(ctx context.Context, deviceID string) []SSHPrivateKey {
+	for _, dir := range s.dirs {
+		s.findPrivateKeys(ctx, dir, deviceID)
+	}
+
+	keys := make([]SSHPrivateKey, 0, len(s.privateKeys))
+	for path, key := range s.privateKeys {
+		keys = append(keys, SSHPrivateKey{
+			Path: path,
+			Key:  key,
+		})
+	}
+	return keys
+}
+
+// SSHPrivateKey represents an SSH private key found during the scan.
+type SSHPrivateKey struct {
+	// Path is the absolute path to the private key file.
+	Path string
+	// Key is the private key representation.
+	Key *accessgraphsecretsv1pb.PrivateKey
+}
+
+// findPrivateKeys walks through all files in a directory and its subdirectories
+// and checks if they are SSH private keys.
+func (s *Scanner) findPrivateKeys(ctx context.Context, root, deviceID string) {
+	logger := s.log.With("dir", root)
+
+	err := filepath.Walk(root, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+		if info.IsDir() {
+			return nil
+		}
+		switch fileData, isKey, err := readFileIfSSHPrivateKey(path); {
+		case err != nil:
+			logger.DebugContext(ctx, "error reading file", "path", path, "error", err)
+		case isKey:
+			key, err := extractSSHKey(ctx, path, deviceID, fileData)
+			if err != nil {
+				logger.DebugContext(ctx, "error extracting private key", "path", path, "error", err)
+			} else {
+				s.privateKeys[path] = key
+			}
+		}
+		return nil
+	})
+
+	if err != nil {
+		logger.WarnContext(ctx, "error walking directory", "root", root, "error", err)
+	}
+}
+
+var (
+	supportedPrivateKeyHeaders = [][]byte{
+		[]byte("RSA PRIVATE KEY"),
+		[]byte("PRIVATE KEY"),
+		[]byte("EC PRIVATE KEY"),
+		[]byte("DSA PRIVATE KEY"),
+		[]byte("OPENSSH PRIVATE KEY"),
+	}
+)
+
+// readFileIfSSHPrivateKey checks if a file is an OpenSSH private key
+func readFileIfSSHPrivateKey(filePath string) ([]byte, bool, error) {
+	file, err := os.Open(filePath)
+	if err != nil {
+		return nil, false, err
+	}
+	defer file.Close()
+
+	// read the first 150 bytes of the file to check if it's an OpenSSH private key.
+	var buf [150]byte
+	n, err := file.Read(buf[:])
+	if errors.Is(err, io.EOF) || n < len(buf) {
+		return nil, false, nil
+	} else if err != nil {
+		return nil, false, trace.Wrap(err, "failed to read file")
+	}
+
+	isPrivateKey := false
+	for _, header := range supportedPrivateKeyHeaders {
+		if bytes.Contains(buf[:], header) {
```

Comment:
> Should we drop the smarts, parse the keys straight away, and go from there?

Actually ignore this part, this could work if ParsePrivateKey accepted a reader, but sadly it does not.